### PR TITLE
Add pill variant to button

### PR DESCRIPTION
**Pull request by yomed**
## Description

- add `pill` variant to the `button`
- update docs

## Context

Pills are very similar to buttons, with a few small differences. So I think it makes sense to keep it within the button component.

A few things I included that I wasn't sure about:
1) I included `margin-left` for adjacent pills. Although Skin is agnostic of the surroundings of the pill, the common use case is to have multiple pills in a row. In which case, the `margin-left` makes sense (and is part of the design spec).
2) I included the CSS multi-line truncation which is available in ~85% of browsers (https://caniuse.com/#feat=css-line-clamp), but is also known to be a bit buggy. We could leave this out of Skin, and just have apps be responsible for handling truncation themselves (or through eBayUI).

## References

Specs: https://ebay.invisionapp.com/share/CWN9X1JF6MA#/screens/293357159
Closes: #69

## Screenshots

screen shot 2018-09-05 at 3 51 35 pm

### CODE REVIEW

**Comment by msendlakowski** [comment]
I think the 2 assumptions make sense.  But we should add a caution in the documentation saying that the truncation isn't completely supported and maybe suggest a JS implementation that can be used.  Or point them to ebayUI once that's complete.

**Comment by msendlakowski** [comment]
Single long words expand outside of the standard pill.

screen shot 2018-09-06 at 1 22 57 pm

We should add overflow hidden.  Its not perfect, but at least it will keep the content contained.  They can add `word-break: break-all` when they know it will be a problem.

**Comment by msendlakowski** [comment]
@yomed I know you're busy on search, but could you give your input on these comments

**Comment by msendlakowski** [comment]
@seangates any concerns?

**Comment by yomed** [comment]
It's broken both with and without `overflow: hidden`, so I'm not sure. I'm inclined to leave it as-is so it's a little more obvious that it's broken, and the devs can figure out where to put the word break on their own.

**Comment by seangates** [comment]
Having done this so many times before and never having a great solution ... we need to do the ellipsis in the component. It's just not reliable through CSS at all. That leaves `overflow: hidden` for Skin, and the ellipsis logic for the component to figure out.

In another thought: do we progressively enhance the style of the button to allow multiline wrapping with ellipsis when JS is present, and a single line with ellipsis when JS does not run?

**Comment by seangates** [comment]
>They can add `word-break: break-all` when they know it will be a problem.

Also, most of the time they won't know it will be a problem before they do translation stuff. Pills will be universally implemented and tried, with so many text string variations we shouldn't offload this to the developers if at all possible. I think we're aiming for a trap if we do that.

**Comment by msendlakowski** [comment]
@seangates So you're saying remove the `line-clamp` and add `overflow: hidden`.  And then do the full implementation in the component using nothing but JS?

**Comment by msendlakowski** [comment]
@seangates @yomed   Yeah, I'm inclined to just remove line-clamp from Skin then.  And not even add `overflow:hidden`.  Since we're not providing a full solution we should make it obvious to applications when it breaks.
What do you think?

**Comment by yomed** [comment]
@msendlakowski Sounds good, that seems consistent.

**Comment by msendlakowski** [comment]
@yomed  This is minor, but it looks like the text is slightly off-center vertically.  A pixel or 2 too high.

screen shot 2018-09-14 at 6 28 58 pm

Also, can you add a note explaining the lack of truncation support and adding a link in the shortcut section here:

screen shot 2018-09-14 at 6 08 56 pm

**Comment by msendlakowski** [comment]
I'll get design input on it.  

## Current
screen shot 2018-09-17 at 10 45 02 am

## With 2px of top padding (and -2px of top margin on btn__cell)
screen shot 2018-09-17 at 10 53 47 am

**Comment by msendlakowski** [comment]
I requested input from Joseph and he sent it to Greg.  Input forthcoming, I think.

**Comment by gbeckpdx** [comment]
Hmm, @msendlakowski, I'd argue the adjusted changes look 2px too low to me now. Do both single and double lined pills appear high to you? I think we'd leave them as is. However what does the code say? Are they truly centered?

**Comment by msendlakowski** [comment]
Using a ruler tool the "Current" is 2px too high, from my measurements.  But its all about perception.  If you're OK with the "Current" then we'll leave it.

**Comment by gbeckpdx** [comment]
Weird. Haha, yeah, must just be a strange optical illusion but thank you for checking! I think it's small enough that we can just go with the original. I definitely notice the adjusted seeming low now...even though it's truly centered. Let's leave for now. I'll be taking a pills v2 pass here soon too, so I'll make sure it's exactly what we want. Thanks!

File: dist/button/ds6/button.css [hunk 1]
```diff
@@ -204,6 +204,30 @@ a.cta-btn--purple:not([href]) {
...
+.cta-btn--pill > .cta-btn__cell > span,
+.fake-btn--pill > .fake-btn__cell > span {
+  max-height: 36px;
+  white-space: normal;
+}
+.btn--pill + .btn--pill,
+.expand-btn--pill + .expand-btn--pill,
+.cta-btn--pill + .cta-btn--pill,
+.fake-btn--pill + .fake-btn--pill {
+  margin-left: 8px;
```
Comment (msendlakowski) [inline dist/button/ds6/button.css]: @yomed @ianmcburnie  What are the chances that different kinds of pills will be used in the same pill set?  This `margin-left` is only applied if the same pill types are adjacent.  If we have a `btn--pill` followed by a `fake-btn--pill` then the margin isn't applied.

What about using:
```
.btn + .btn--pill
.btn + .expand-btn--pill
etc.
```

Comment (msendlakowski) [reply dist/button/ds6/button.css]: screen shot 2018-09-19 at 5 28 08 pm

Comment (ianmcburnie) [reply dist/button/ds6/button.css]: What about:

```css
[class*="btn--pill"] + [class*="btn--pill"] {
  margin-left: 8px;
}
```

Would that capture all cases? Please bear in mind I've been away 6 weeks ;-)

**Comment by gbeckpdx** [comment]
@ianmcburnie at this juncture, we've not been building out actual DS4 components in our libraries for designers, so I'm not sure of how we're handling this in eBayUI. How are you thinking about this in general? It's probably a larger discussion we need to have as a team, but I think it depends on how far away people are from making the switch to the new DS evolution...

**Comment by ianmcburnie** [comment]
@gbeckpdx Right now if a DS4 page attempts to use the pill variant of the button component, they will receive an unstyled (or possibly broken styled) button. We could change it so that they receive a fully styled ds6 button, but obviously that's going to look a little odd alongside ds4 stuff on the same page. Personally I'd rather have *something* in place for DS4 if it's low hanging fruit. For example, keeping the same look and feel, but using ds4 colour palette. I guess from your point of view though, you might feel that philosophically the pill module does not and should not belong in DS4.

**Comment by ianmcburnie** [comment]
@msendlakowski @yomed I think we should revisit how wrapping works for all buttons. See ticket https://github.com/eBay/skin/issues/391. If we fix this in button, then we fix it in pill. I recognize that this probably adds a lot more days to this task, and you probably don't have the bandwidth, but I'm sure somebody else can pick it up if we think this is a good idea.

File: dist/button/ds6/button.css (hunk 1 above)
Comment (yomed) [reply dist/button/ds6/button.css]: Sure, looks fine

**Comment by gbeckpdx** [comment]
@ianmcburnie DS4 and DS6 should stay separate when teams ask for either. If they want DS4, do we have components for the old stuff or would that be net-new work?

**Comment by msendlakowski** [comment]
@ianmcburnie  I added notes in that issue you tagged.  I don't know if there is a great solution, but its worth looking into.  Right now the button overflow situation is pretty bad.

**Comment by ianmcburnie** [comment]
@gbeckpdx  If they want DS4, then no, right now we do not have a ds4 component for pills.

**Comment by seangates** [comment]
@ianmcburnie That's very close to what I proposed with having the truncation happen within the JS. Essentially, it's the same idea.

**Comment by seangates** [comment]
e.g. Handle truncating by detecting overflow through JS.

**Comment by ianmcburnie** [comment]
Ok, cool. Just checking we are on the same page as I didn't see any mention of a modifier.

**Comment by seangates** [comment]
Adding the modifier is JS, though. 🤷‍♂️

**Comment by seangates** [comment]
🤷‍♂️

**Comment by msendlakowski** [comment]
One thing to consider... the pill is restricted by the DS to 2 lines, and has a fixed with at that height.  Will we restrict the other buttons to 2 lines?  If so, do we truncate then?  If not, then we still need custom behavior for the pill.

**Comment by gbeckpdx** [comment]
> @gbeckpdx If they want DS4, then no, right now we do not have a ds4 component for pills.

I don't have the answer, but I suppose the question would be are we as a team going to support DS4? Might just be diminishing returns to flesh out a whole separate library that's phasing out eh? I'd imagine you guys don't want to double your work load eh? ;) Not sure what the right answer is for this. I'd defer to Joseph.

**Comment by ianmcburnie** [comment]
After discussion with Product Designs System team, it seems likely that DS4 pages will try and incorporate pills into their pages. Therefore as a follow up task we should add DS4 variant of pill module (i.e. exactly the same, just using ds4 colours).

**Comment by ianmcburnie** [comment]
Also note that we are going to put the brakes on the pill skin module and core component, while we figure out text wrapping as a whole in buttons (a task for next sprint). We might as well get this right before putting it out.

**Comment by msendlakowski** [comment]
When looking at wrapping, keep in mind the a11y issue of low vision user's with larger font.  Right now the pill/button size doesn't grow with the text size, so a lot of info is lost.  Truncation isn't applied in this example yet, but you can see the issue in the third button:

screen shot 2018-09-24 at 9 57 48 am


screen shot 2018-09-24 at 9 45 25 am

**Comment by gbeckpdx** [comment]
Yeah just so you guys know too, I'm about to start a v2 on Pills that address a bunch on these gaps. I'll keep you updated on what's going on every week, but if you could let me know the majority of problems we're hearing about, that would be great to gather now. Thanks!

**Comment by ianmcburnie** [comment]
Going to retarget this for 7.0.0 branch. This PR will need rebasing with the new button updates once they are in. The new way we are doing buttons (PR #401) will most likely have a knock on effect for pills and hopefully make the text wrapping situation a little easier.

@seangates After the button changes are merged to 7.0.0 branch, we will need to revisit this PR and see what still makes sense.

**Comment by seangates** [comment]
Closing as we have removed the base class has changed a lot. Will create a new PR.

**Pull request closed without merging**